### ISSUE TITLE
OMXPlayer hangs/corrupts heap on corrupted media

### ISSUE THREAD

**Issue by xKaiser** [popcornmix/omxplayer#630]
When playing and killings corrupted H264 videos in a loop with the OMXPlayer, the GPU memory heap becomes corrupted: "sudo vcdbg reloc" will report "heap corruption detected". Expected behavior is to reject the file, but I'm not sure what's possible.

A corrupted media file that triggers this is issue is created by transferring to an FTP with Filezilla in ASCII mode, and then to the pi in Binary mode. 

Steps to reproduce are:
While heap not corrupted
-Play corrupted media with omxplayer (omxplayer -o hdmi corruptedMedia.mp4)
-This will hang and never complete, so kill it (ctrl + C or pkill from another terminal is sufficient)

Example file that triggers this issue:
https://drive.google.com/open?id=1T6NNIdsXQyGe-O5W1ps1GvrMK6RNndFM

**Comment by nielsbuus** [comment]
@xKaiser did you ever find a solution to this? I am facing the exact same problem. And I notice that sometimes when I terminate a valid file, `vcdbg reloc` will still report `heap corruption detected`.

    root@raspberrypi:~# vcdbg reloc

    Relocatable heap version 4 found at 0x38000000
    total space allocated is 108M, with 108M relocatable, 0 legacy and 0 offline
    0 legacy blocks of size 2359296

    free list at 0x3e512f60
    101M free memory in 2 free block(s)
    largest free block is 101M bytes

    0x38000000: free 101M
    [ 120] 0x3e512b00: used  576 (refcount 1 lock count 0, size      512, align    4, data 0x3e512b20, d0rual) 'ILCS VC buffer pool'
    [  73] 0x3e512d40: used  576 (refcount 1 lock count 0, size      512, align    4, data 0x3e512d60, d0rual) 'ILCS VC buffer pool'
    0x3e512d40: corrupt trailer (space 576 != 577)
    [   3] 0x3e512f80: used 6.9M (refcount 1 lock count 8, size  7237632, align 4096, data 0x3e513000, d1rual) 'ARM FB'
    [   2] 0x3ebfafa0: used  16K (refcount 1 lock count 0, size    16384, align   32, data 0x3ebfafc0, d0ruAl) 'audioplus_tmp_buf'
    [   1] 0x3ebfefe0: used 4.0K (refcount 1 lock count 0, size        0, align 4096, data 0x3ebff000, d1rual) 'camera fast alloc arena'
    heap corruption detected
    small allocs not requested

**Comment by xKaiser** [comment]
Kind of, I switched to using a rock64 and gstreamer. Huge pain in the ass to switch platforms, but completely worth it, as it's cheaper, has more juice, and more throughput.

As far as I know, you cannot safely kill omxplayer, so until the issue is addressed, there's nothing more you can do. I even looked into resetting the memory for the video chip, but the only mechanism for that is a soft reset...

So in short, if you're looking for reliability, I'd run from the pi.

-------- Original Message --------

**Comment by nielsbuus** [comment]
Thanks, although not the answer I was hoping for. 😄 

It sounds like we are trying to solve the same problem. Can I ask you what sort of product you are building? Do you have any blog, description or (if open source) source code that I can look at? I'm particularly interested in knowing just how much pain it was, and the potential shortcomings or challenges you encountered along the way.

If you don't like to discuss it within this issue, feel free to reach me at nielsbuus@gmail.com

**Comment by xKaiser** [comment]
It was a system that plays media in response to sensor data, and downloaded assets from the internet (from an FTP where sometimes people don't transfer in binary mode and corrupt the videos).

Reliability of the pi is lacking when video is involved, so switching platforms is in your best interest. There's tons of guides on gstreamer and other tools (mpv) to handle media on other devices.

-------- Original Message --------
